 Allow some rounding on SO_TIMEOUT values

Kernels can round the value in use depending on the HZ setting.
This change allows up to 10ms of slop between the SO_TIMEOUT
value set and the one returned.

diff --git a/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/DatagramSocketTest.java b/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/DatagramSocketTest.java
index d9f3d91..33e0223 100644
--- a/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/DatagramSocketTest.java
+++ b/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/DatagramSocketTest.java
@@ -390,8 +390,11 @@
 
     public void test_getSoTimeout() throws Exception {
         DatagramSocket ds = new DatagramSocket();
-        ds.setSoTimeout(100);
-        assertEquals("Returned incorrect timeout", 100, ds.getSoTimeout());
+        final int timeoutSet = 100;
+        ds.setSoTimeout(timeoutSet);
+        int actualTimeout = ds.getSoTimeout();
+        // The kernel can round the requested value based on the HZ setting. We allow up to 10ms.
+        assertTrue("Returned incorrect timeout", Math.abs(actualTimeout - timeoutSet) <= 10);
     }
 
     static final class TestDatagramSocketImpl extends DatagramSocketImpl {
diff --git a/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/ServerSocketTest.java b/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/ServerSocketTest.java
index d7ef745..f42ba24 100644
--- a/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/ServerSocketTest.java
+++ b/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/ServerSocketTest.java
@@ -296,9 +296,12 @@
      */
     public void test_getSoTimeout() throws IOException {
         s = new ServerSocket(0);
+        final int timeoutSet = 100;
         try {
-            s.setSoTimeout(100);
-            assertEquals("Returned incorrect sotimeout", 100, s.getSoTimeout());
+            s.setSoTimeout(timeoutSet);
+            int actualTimeout = s.getSoTimeout();
+            // The kernel can round the requested value based on the HZ setting. We allow up to 10ms.
+            assertTrue("Returned incorrect sotimeout", Math.abs(timeoutSet - actualTimeout) <= 10);
         } finally {
             s.close();
         }
@@ -340,12 +343,15 @@
      */
     public void test_setSoTimeoutI() throws IOException {
         // Timeout should trigger and throw InterruptedIOException
+        final int timeoutSet = 100;
         try {
             s = new ServerSocket(0);
-            s.setSoTimeout(100);
+            s.setSoTimeout(timeoutSet);
             s.accept();
         } catch (InterruptedIOException e) {
-            assertEquals("Set incorrect sotimeout", 100, s.getSoTimeout());
+            int actualSoTimeout = s.getSoTimeout();
+            // The kernel can round the requested value based on the HZ setting. We allow up to 10ms.
+            assertTrue("Set incorrect sotimeout", Math.abs(timeoutSet - actualSoTimeout) <= 10);
             return;
         }
 
diff --git a/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/SocketTest.java b/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/SocketTest.java
index 4df92e2..0915a27 100644
--- a/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/SocketTest.java
+++ b/harmony-tests/src/test/java/org/apache/harmony/tests/java/net/SocketTest.java
@@ -967,8 +967,11 @@
     public void test_getSoTimeout() throws Exception {
         ServerSocket server = new ServerSocket(0);
         Socket client = new Socket(InetAddress.getLocalHost(), server.getLocalPort());
-        client.setSoTimeout(100);
-        assertEquals("Returned incorrect sotimeout", 100, client.getSoTimeout());
+        final int timeoutSet = 100;
+        client.setSoTimeout(timeoutSet);
+        int actualTimeout = client.getSoTimeout();
+        // The kernel can round the requested value based on the HZ setting. We allow up to 10ms.
+        assertTrue("Returned incorrect sotimeout", Math.abs(timeoutSet - actualTimeout) <= 10);
         client.close();
         server.close();
     }
@@ -1492,8 +1495,11 @@
     public void test_setSoTimeoutI() throws Exception {
         ServerSocket server = new ServerSocket(0);
         Socket client = new Socket(InetAddress.getLocalHost(), server.getLocalPort());
-        client.setSoTimeout(100);
-        assertEquals("Set incorrect sotimeout", 100, client.getSoTimeout());
+        final int timeoutSet = 100;
+        client.setSoTimeout(timeoutSet);
+        int actualTimeout = client.getSoTimeout();
+        // The kernel can round the requested value based on the HZ setting. We allow up to 10ms.
+        assertTrue("Set incorrect sotimeout", Math.abs(timeoutSet - actualTimeout) <= 10);
         client.close();
         server.close();
     }
diff --git a/luni/src/test/java/libcore/java/net/OldServerSocketTest.java b/luni/src/test/java/libcore/java/net/OldServerSocketTest.java
index 6518897..8588144 100644
--- a/luni/src/test/java/libcore/java/net/OldServerSocketTest.java
+++ b/luni/src/test/java/libcore/java/net/OldServerSocketTest.java
@@ -238,11 +238,11 @@
     public void test_getSoTimeout_setSoTimeout() throws Exception {
         // TODO: a useful test would check that setSoTimeout actually causes timeouts!
         ServerSocket s = new ServerSocket();
-        s.setSoTimeout(1500);
-        int ms = s.getSoTimeout();
-        if (ms < 1500-10 || ms > 1500+10) {
-            fail("suspicious timeout: " + ms);
-        }
+        final int timeoutSet = 1500;
+        s.setSoTimeout(timeoutSet);
+        int actualTimeout = s.getSoTimeout();
+        // The kernel can round the requested value based on the HZ setting. We allow up to 10ms.
+        assertTrue("suspicious timeout: " + actualTimeout, Math.abs(actualTimeout - timeoutSet) <= 10);
         s.close();
         try {
             s.getSoTimeout();
diff --git a/luni/src/test/java/libcore/java/net/OldSocketTest.java b/luni/src/test/java/libcore/java/net/OldSocketTest.java
index ded5802..4254516 100644
--- a/luni/src/test/java/libcore/java/net/OldSocketTest.java
+++ b/luni/src/test/java/libcore/java/net/OldSocketTest.java
@@ -389,9 +389,7 @@
         Socket s = new Socket();
         s.setSoTimeout(1500);
         int ms = s.getSoTimeout();
-        if (ms < 1500-10 || ms > 1500+10) {
-            fail("suspicious timeout: " + ms);
-        }
+        assertTrue("suspicious timeout: " + ms, Math.abs(ms - 1500) <= 10);
         s.close();
         try {
             s.getSoTimeout();
@@ -1384,10 +1382,10 @@
         connector.start();
         theSocket.setSoTimeout(1000);
         Thread.sleep(10);
-        assertEquals("Socket option not set during connect: 10 ", 1000, theSocket.getSoTimeout());
+        assertTrue("Socket option not set during connect: 10 ", Math.abs(1000 - theSocket.getSoTimeout()) <= 10);
         Thread.sleep(50);
         theSocket.setSoTimeout(2000);
-        assertEquals("Socket option not set during connect: 50 ", 2000, theSocket.getSoTimeout());
+        assertTrue("Socket option not set during connect: 50 ", Math.abs(2000 - theSocket.getSoTimeout()) <= 10);
         Thread.sleep(5000);
         theSocket.close();
 
diff --git a/luni/src/test/java/libcore/javax/net/ssl/SSLSocketTest.java b/luni/src/test/java/libcore/javax/net/ssl/SSLSocketTest.java
index 11dfb3d..274712f 100644
--- a/luni/src/test/java/libcore/javax/net/ssl/SSLSocketTest.java
+++ b/luni/src/test/java/libcore/javax/net/ssl/SSLSocketTest.java
@@ -1206,8 +1206,9 @@
         // setting wrapper sets underlying and ...
         int expectedTimeoutMillis = 1000;  // 10 was too small because it was affected by rounding
         wrapping.setSoTimeout(expectedTimeoutMillis);
-        assertEquals(expectedTimeoutMillis, wrapping.getSoTimeout());
-        assertEquals(expectedTimeoutMillis, underlying.getSoTimeout());
+        // The kernel can round the requested value based on the HZ setting. We allow up to 10ms.
+        assertTrue(Math.abs(expectedTimeoutMillis - wrapping.getSoTimeout()) <= 10);
+        assertTrue(Math.abs(expectedTimeoutMillis - underlying.getSoTimeout()) <= 10);
 
         // ... getting wrapper inspects underlying
         underlying.setSoTimeout(0);
